#### Question 1

In [1]:
import bentoml

In [2]:
bentoml.__version__

'1.0.7'

#### Question 2

In [4]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction import DictVectorizer
from sklearn.metrics import roc_auc_score
from sklearn.ensemble import RandomForestClassifier

In [5]:
data = 'https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-06-trees/CreditScoring.csv'
df = pd.read_csv(data)

In [6]:
df.columns = df.columns.str.lower()

status_values = {
    1: 'ok',
    2: 'default',
    0: 'unk'
}

df.status = df.status.map(status_values)

home_values = {
    1: 'rent',
    2: 'owner',
    3: 'private',
    4: 'ignore',
    5: 'parents',
    6: 'other',
    0: 'unk'
}

df.home = df.home.map(home_values)

marital_values = {
    1: 'single',
    2: 'married',
    3: 'widow',
    4: 'separated',
    5: 'divorced',
    0: 'unk'
}

df.marital = df.marital.map(marital_values)

records_values = {
    1: 'no',
    2: 'yes',
    0: 'unk'
}

df.records = df.records.map(records_values)

job_values = {
    1: 'fixed',
    2: 'partime',
    3: 'freelance',
    4: 'others',
    0: 'unk'
}

df.job = df.job.map(job_values)

for c in ['income', 'assets', 'debt']:
    df[c] = df[c].replace(to_replace=99999999, value=np.nan)

df = df[df.status != 'unk'].reset_index(drop=True)

In [7]:
df_train, df_test = train_test_split(df, test_size=0.2, random_state=11)

df_train = df_train.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

y_train = (df_train.status == 'default').astype('int').values
y_test = (df_test.status == 'default').astype('int').values

del df_train['status']
del df_test['status']

In [8]:
dv = DictVectorizer(sparse=False)

train_dicts = df_train.fillna(0).to_dict(orient='records')
X_train = dv.fit_transform(train_dicts)

test_dicts = df_test.fillna(0).to_dict(orient='records')
X_test = dv.transform(test_dicts)

In [9]:
rf = RandomForestClassifier(n_estimators=200,
                            max_depth=10,
                            min_samples_leaf=3,
                            random_state=1)
rf.fit(X_train, y_train)

RandomForestClassifier(max_depth=10, min_samples_leaf=3, n_estimators=200,
                       random_state=1)

In [10]:
bentoml.sklearn.save_model(
    'credit_risk_model',
    rf,
    custom_objects={'dict_vectorizer': dv},
)

Model(tag="credit_risk_model:pbubx22qpkrfk5cm", path="C:\Users\Alex\bentoml\models\credit_risk_model\pbubx22qpkrfk5cm\")

In [2]:
!bentoml models list


 Tag                          Module           Size        Creation Time       
 credit_risk_rf_model:bhwa4…  bentoml.sklearn  5.53 MiB    2022-10-20 16:42:28 
 credit_risk_model:ml3trl2q…  bentoml.xgboost  197.77 KiB  2022-10-20 16:30:38 
 mlzoomcamp_homework:qtzdz3…  bentoml.sklearn  5.79 KiB    2022-10-13 23:42:14 


credit_risk_rf_model size is 5.53 MB 

#### Question 3

In [ ]:
from pydantic import BaseModel

class UserProfile(BaseModel):
    name: str
    age: int
    country: str
    rating: float

#### Question 4

In [ ]:
!curl -O https://s3.us-west-2.amazonaws.com/bentoml.com/mlzoomcamp/coolmodel.bentomodel
!bentoml models import coolmodel.bentomodel

In [3]:
!bentoml models list

 Tag                          Module           Size        Creation Time       
 credit_risk_rf_model:bhwa4…  bentoml.sklearn  5.53 MiB    2022-10-20 16:42:28 
 credit_risk_model:ml3trl2q…  bentoml.xgboost  197.77 KiB  2022-10-20 16:30:38 
 mlzoomcamp_homework:qtzdz3…  bentoml.sklearn  5.79 KiB    2022-10-13 23:42:14 


In [7]:
%pycat C:\Users\Alex\bentoml\models\mlzoomcamp_homework\qtzdz3slg6mwwdu5\model.yaml

name: mlzoomcamp_homework
version: qtzdz3slg6mwwdu5
module: bentoml.sklearn
labels: {}
options: {}
metadata: {}
context:
  framework_name: sklearn
  framework_versions:
    scikit-learn: 1.1.1
  bentoml_version: 1.0.7
  python_version: 3.9.12
signatures:
  predict:
    batchable: false
api_version: v1
creation_time: '2022-10-13T20:42:14.411084+00:00'


#### Question 5
answer: 1

#### Question 6